### PR TITLE
Native form to patient portal form

### PR DESCRIPTION
**Pull request by bradymiller**
@sjpadgett , Have the main plumbing in (can save the form and modify the form form the patient portal).  Plan:
1. Get it working well in portal and in the dashboard where it is saved to chart as pdf etc. **DONE**
2. Streamline the process of how a "patient portal compliant" form can be placed in template (will potentially do this from the Forms->Adminstration gui for forms that are "patient portal compliant") - **DONE - (patient portal compliant forms get a neat icon next to their names in the gui and the template is placed into the document_templates sql table when the form is registered)**
3. Encapsulate things as much as possible to make it as easy as possible for users to make their forms "patient portal compliant" **DONE**

btw, the first commit in this PR is security related bug fixes and then the second commit is where the magic is starting :)

Up For Grabs demo (includes the standard demo data) is at:
https://www.open-emr.org/wiki/index.php/Development_Demo#Epsilon_-_Up_For_Grabs_Demo

### CODE REVIEW

File: portal/patient/scripts/app/onsitedocuments.js [hunk 1]
```diff
@@ -656,12 +656,20 @@ var page = {
...
+                                            "?isPortal=" + encodeURIComponent(isPortal ? 1 : 0) +
+                                            "&formOrigin=" + encodeURIComponent(page.formOrigin) +
+                                            "&formname=" + encodeURIComponent(page.lbfFormName) + "&id=0";
+                                    } else {
+                                        // iframe from template directive {EncounterDocument:xxxxx} for a native form
+                                        url = webRoot + "/interface/forms/" + encodeURIComponent(page.lbfFormName) + "/new.php" +
+                                            "?isPortal=" + encodeURIComponent(isPortal ? 1 : 0) +
+                                            "&formOrigin=" + encodeURIComponent(page.formOrigin) +
+                                            "&formname=" + encodeURIComponent(page.lbfFormName) + "&id=0";
+                                    }
```
Comment (sjpadgett) [inline portal/patient/scripts/app/onsitedocuments.js:672]: This is exactly the approach I'd taken. Besides the audit, we may want to send a notification to user for form completion.
This is something I want to do a better job at doing meaning, notifications in portal. Not only user but SMS/email to patient via event.

Looking good so far and let me know if I can help.

**Comment by bradymiller** [comment]
Up For Grabs demo (includes the standard demo data) is at:
https://www.open-emr.org/wiki/index.php/Development_Demo#Epsilon_-_Up_For_Grabs_Demo

File: interface/forms/sdoh/new.php [hunk 2]
```diff
@@ -156,6 +184,25 @@ function CalculateTotal() {
...
+                    }
+                    if (e.data.submitForm === true) {
+                        e.preventDefault();
+                        document.forms[0].submit();
+                    }
+                    <?php if (($mode == "update") && $patientPortalOther) { ?>
+                        parent.postMessage({formid:<?php echo js_escape($_GET['id']); ?>}, window.location.origin);
+                    <?php } ?>
+                });
+            });
```
Comment (sjpadgett) [inline interface/forms/sdoh/new.php:203]: Glad you caught this. 
Although, all these modifications that are necessary to make an encounter form compliant with portal seems to me cumbersome for hope others can follow the pattern.

I've been thinking about trying to come up with a canned bootstrap in forms for doing the same.

**Comment by bradymiller** [comment]
am working on a class that turns all these code blocks into 1 liners (and
dropping the .tpl file). should have it in the pr by tonight.

**Comment by bradymiller** [comment]
This is testing well and ready for code review .

@sjpadgett , tried to make the required changes as minimal as possible. Can see example of the changes made in the sdoh form on the following commit: https://github.com/openemr/openemr/pull/5530/commits/55ef2d10cc9479378f26bdd634c9a138698b14df

**Review by sjpadgett — APPROVED** [review]
Nice Brady. I don't think it can be made any easier and still stay secure and appropriate for patients.
I don't think it'd be to hard to convert existing forms if wanted.

**Comment by bradymiller** [comment]
@charjmiller , Your form can now be completed in the patient portal :) Also updated wiki page here: https://www.open-emr.org/wiki/index.php/Social_Screening_Tool

@kofiav , check out this commit for how to make a encounter form to be patient portal compliant: https://github.com/openemr/openemr/commit/3f0d87cd25464bc01d1a923e1868aabef392e38b

@ruthkonyn , figured you would be interested in this for your phq9 and gad7 forms

**Comment by charjmiller** [comment]
Looks great @bradymiller !  I'll work on adding this documentation onto the User Guide.

**Pull request merged**